src/Storages/MergeTree/DataPartsExchange.h

```diff
@@ -29,16 +35,27 @@ class Service final : public InterserverIOEndpoint
     std::string getId(const std::string & node_id) const override;
     void processQuery(const Poco::Net::HTMLForm & params, ReadBuffer & body, WriteBuffer & out, Poco::Net::HTTPServerResponse & response) override;
 
+    void setZooKeeper(const zkutil::ZooKeeperPtr & zookeeper_, const String & zookeeper_path_, const String & replica_name_) override
+    {
+        zookeeper = zookeeper_;
+        zookeeper_path = zookeeper_path_;
+        replica_name = replica_name_;
+    }
+
 private:
     MergeTreeData::DataPartPtr findPart(const String & name);
     void sendPartFromMemory(const MergeTreeData::DataPartPtr & part, WriteBuffer & out);
     void sendPartFromDisk(const MergeTreeData::DataPartPtr & part, WriteBuffer & out, int client_protocol_version);
+    void sendPartS3Metadata(const MergeTreeData::DataPartPtr & part, WriteBuffer & out);
 
 private:
     /// StorageReplicatedMergeTree::shutdown() waits for all parts exchange handlers to finish,
     /// so Service will never access dangling reference to storage
     MergeTreeData & data;
     Poco::Logger * log;
+    zkutil::ZooKeeperPtr zookeeper;
```

Comment:
I'm sorry, the GitHub review interface is not so good but where we use it? I cannot find :(